BayesianEstimation/AdaptiveMetropolisWithinGibbs.cs

```diff
@@ -0,0 +1,285 @@
+    /// function. The rest of the algorithm remains the same (the Metropolis within Gibbs sampler samples from the distributions 
+    /// you define in order to fit the distribution you want to your data). See FlashLfqBayesianEstimator.cs for an example of defining
+    /// a different posterior probability function.
+    /// 
+    /// See:
+    /// Bayesian Estimation Supersedes the t Test. Kruschke, J. K., Journal of Experimental Psychology, 2013.
+    /// 
+    /// This code is adapted from Javascript code taken from http://sumsar.net/best_online/ . This webpage is a useful
+    /// interactive example of J. Kruschke's Bayesian equivalent of a t-test.
+    /// </summary>
+    public class AdaptiveMetropolisWithinGibbs
+    {
+        public List<double[]> markovChain { get; private set; }
+        private double[] logSd;
+        private double[] currentState;
+        private double[] acceptanceCount;
+        private int numParams;
+        private int batchCount;
+        private int batchSize;
+        private MersenneTwister random;
+        private List<double>[] Data;
+        private bool isTwoSample;
+        private List<Tuple<double, double>>[] acceptableParameterRanges;
+
+        // these are the parameters of the prior probability distributions
+        private double meanMu; // mean of prior prob mean normal distribution
+        private double sdMu; // sd of prior prob mean normal distribution
+        private double sigmaLow; // lower bound of prior prob sd uniform distribution
+        private double sigmaHigh; // upper bound of prior prob sd uniform distribution
+        private double exp; // exponent of prior prob exponential distribution
+
+        /// <summary>
+        /// Construct the adaptive Metropolis within Gibbs sampler. Leave data2 null for one-sample data.
+        /// </summary>
+        public AdaptiveMetropolisWithinGibbs(List<double> data1, List<double> data2, int batch_size = 3, int? seed = null)
+        {
+            if (seed != null)
+            {
+                random = new MersenneTwister(seed.Value, true);
+            }
+            else
+            {
+                random = new MersenneTwister(true);
+            }
+
+            batchCount = 0;
+            this.batchSize = batch_size;
+            isTwoSample = data2 != null;
+
+            if (isTwoSample)
+            {
+                Data = new List<double>[] { data1, data2 };
+            }
+            else
+            {
+                Data = new List<double>[] { data1 };
+            }
+            
+            SetUpAcceptableParameterRanges();
+            SetUpPriorProbabilityDistributions();
+            SetUpPosteriorFunction(isTwoSample);
+        }
+
+        /// <summary>
+        /// Burn in and then sample the MCMC chain.
+        /// </summary>
+        public void Run(int burnin = 20000, int n = 20000)
```

Comment:
what's special about 20000. i guess depending on set size, particularly small sets, you get redundancy?

Reply:
nothing really - it's generally enough iterations to converge on a reasonable answer in most cases. more iterations of the MCMC algorithm produce more accurate answers, for the point estimate and uncertainty for each parameter